crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,192 @@
+
+def compare_crashes(pings, start_date, end_date, comparable_dimensions, dimension_names):
+    """Returns a PairRDD where keys are user configurations and values are Numpy arrays of the form [usage hours, main process crashes, content process crashes, plugin crashes]"""
+    ping_properties = get_pings_properties(pings, comparable_dimensions + [
+        "meta/submissionDate",
+        "creationDate",
+        "payload/info/subsessionLength",
+        "meta/docType",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/content",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/plugin",
+        "payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/gmplugin",
+    ], with_processes=True)
+    def is_valid(ping): # sanity check to make sure the ping is actually usable for our purposes
+        submission_date = ping["meta/submissionDate"]
+        if not isinstance(submission_date, str) and not isinstance(submission_date, unicode):
+            return False
+        activity_date = ping["creationDate"]
+        if not isinstance(activity_date, str) and not isinstance(activity_date, unicode):
+            return False
+        subsession_length = ping["payload/info/subsessionLength"]
+        if subsession_length is not None and not isinstance(subsession_length, int) and not isinstance(subsession_length, long):
+            return False
+        return True
+
+    def get_crash_pair(ping): # responsible for normalizing a single ping into a crash pair
+        # we need to parse and normalize the dates here rather than at the aggregates level,
+        # because we need to normalize and get rid of the time portion
+
+        # date the ping was received
+        submission_date = datetime.strptime(ping["meta/submissionDate"], "%Y%m%d").date() # convert the YYYYMMDD format to a real date
+        submission_date = max(start_date, min(end_date, submission_date)) # normalize the submission date if it's out of range
+
+        # date the ping was created on the client
+        try:
+            activity_date = dateutil.parser.parse(ping["creationDate"]).date() # the activity date is the date portion of creationDate
+        except: # malformed date, like how sometimes seconds that are not in [0, 59]
+            activity_date = submission_date # we'll just set it to the submission date, since we can't parse the activity date
+        activity_date = max(submission_date - timedelta(days=7), min(submission_date, activity_date)) # normalize the activity date if it's out of range
+
+        return (
+            # the keys we want to filter based on
+            (submission_date, activity_date) + tuple(ping[key] for key in comparable_dimensions), # all the dimensions we can compare by
+            # the crash values
+            np.array([
+                min(25, (ping["payload/info/subsessionLength"] or 0) / 3600.0), # usage hours, limited to ~25 hours to keep things normalized
```

Comment:
In that case, the upper bound for a subsession length would be 1 hour, which doesn't really cover the actual distribution of subsession lengths.